### PR TITLE
Step 2 - Implementing Functions

### PR DESCRIPTION
**Pull request by michmitz**
**1. What this does:**
* Implements CRUD functions for user class:
   - Get user by id
   - Get all users
   - Create user
   - Delete user
   - Update user
* Creates CustomError class with customized status code and message
* Tests edge cases for each function, such as when no user can be found by id.

**2. Ticket:**
Discussed with Aron.

**3. How to test:**
-All tests are currently passing.

### CODE REVIEW

File: src/index.test.ts [hunk 1]
```diff
@@ -17,15 +43,20 @@ describe('Tests will go here!', () => {
...
-      expect(result).toStrictEqual(null)
-  })
+  it('returns an error if there is no user matching id', () => {
+      expect(() => users.getUserById(10)).toThrow()
+      try {
+           users.getUserById(10)
+      } catch (error: any) {
+          expect(error).toBeInstanceOf(CustomError)
+          expect(error.message).toBe("No user found.")
+          expect(error.status).toEqual(500)
```
Comment (ianandrewson) [inline src/index.test.ts]: I'm no expert in this, but I'm not sure status 500 is the most semantic status you can use here. On the one hand, it is the server running the code to find the user. On the other hand, it's a specific resource that can't be found, likely queried by the client. Aron should probably weigh in on what the best status codes to use here are. Personally, I implemented error codes in the 400 range.

File: src/index.test.ts [hunk 2]
```diff
@@ -55,13 +86,20 @@ describe('Tests will go here!', () => {
     ]) 
   }) 
 
-  it.skip('returns null if no users are found', () => {
-      const result = users.getUserById(10)
+  it('returns CustomError if no users are found', () => {
```
Comment (ianandrewson) [inline src/index.test.ts]: The behavior here would probably be defined by a product owner or someone other than you or I, but there is a semantic argument to be made that when getting all users, returning an empty is totally fine instead of returning an error. Technically, all users were found, there just aren't any. And as long as you return an empty array and not null, any front end implementation shouldn't choke on an empty array.

So the question becomes, is finding no users an error?

File: src/index.test.ts [hunk 3]
```diff
@@ -55,13 +86,20 @@ describe('Tests will go here!', () => {
...
+    const noUsers = new UserAPI([])
+    expect(() => noUsers.getUsers()).toThrow()
 
-      expect(result).toStrictEqual(null)
+    try {
+      noUsers.getUsers()
+    } catch (error: any) {
+        expect(error).toBeInstanceOf(CustomError)
+        expect(error.message).toBe("No users found.")
+        expect(error.status).toEqual(500)
```
Comment (ianandrewson) [inline src/index.test.ts]: A more specific status code can likely be given here as well.

File: src/index.test.ts [hunk 4]
```diff
@@ -78,18 +116,25 @@ describe('Tests will go here!', () => {
...
+    }
 
-      expect(result).toStrictEqual(null)
+    expect(() => users.createUser(newUser)).toThrow()
+      try {
+      users.createUser(newUser)
+    } catch (error: any) {
+        expect(error).toBeInstanceOf(CustomError)
+        expect(error.message).toBe("User with id already exists.")
+        expect(error.status).toEqual(500)
```
Comment (ianandrewson) [inline src/index.test.ts]: Here too.

File: src/index.test.ts [hunk 5]
```diff
@@ -98,15 +143,21 @@ describe('Tests will go here!', () => {
...
-  it.skip('updates a user by id', () => {
+  it('returns CustomError if deleted user id does not exist', () => {
+    expect(() => users.deleteUserById(10)).toThrow()
+    
+    try {
+        users.deleteUserById(10)
+      } catch (error: any) {
+          expect(error).toBeInstanceOf(CustomError)
+          expect(error.message).toBe("No user with that id found.")
+          expect(error.status).toEqual(500)
```
Comment (ianandrewson) [inline src/index.test.ts]: And here.

File: src/index.test.ts [hunk 6]
```diff
@@ -121,14 +172,41 @@ describe('Tests will go here!', () => {
...
+    }
+    
+    expect(() => users.updateUser(1, updatedUser)).toThrow()
+
+    try {
+        users.updateUser(1, updatedUser)
+      } catch (error: any) {
+          expect(error).toBeInstanceOf(CustomError)
+          expect(error.message).toBe("Different user with same id already exists.")
+          expect(error.status).toEqual(500)
```
Comment (ianandrewson) [inline src/index.test.ts]: Yup, and here.

File: src/index.test.ts [hunk 7]
```diff
@@ -121,14 +172,41 @@ describe('Tests will go here!', () => {
...
+    }
+    
+    expect(() => users.updateUser(10, updatedUser)).toThrow()
+
+    try {
+        users.updateUser(10, updatedUser)
+      } catch (error: any) {
+          expect(error).toBeInstanceOf(CustomError)
+          expect(error.message).toBe("No user found by that id.")
+          expect(error.status).toEqual(500)
```
Comment (ianandrewson) [inline src/index.test.ts]: The last one.

File: src/index.ts [hunk 8]
```diff
@@ -7,54 +7,69 @@ export interface User {
...
+    readonly message: string
+}
+
+export class CustomError extends Error {
+    status: number
+    
+    constructor(status: number, message: string) {
+      super(message);
+      this.status = status
+      Object.setPrototypeOf(this, CustomError.prototype)
```
Comment (ianandrewson) [inline src/index.ts:20]: Very curious why you included this. I have no idea if it's right or wrong or anything; this isn't something I did in my implementation and I'm curious as to why you did this.

File: src/index.ts [hunk 9]
```diff
@@ -7,54 +7,69 @@ export interface User {
...
+    }  
 
-    public deleteUserById = (id: number): User | null => {
-        console.log("Create user by id")
-    }
+    public deleteUserById = (id: number): User => {
+        const user = this.users.find(existingUser => existingUser.id === id)
+        
+        if (user) {
+            new UserAPI(this.users.filter(user => user.id !== id))
```
Comment (ianandrewson) [inline src/index.ts]: I still firmly believe there is a way to do this without instantiating a new version of the class.

File: src/index.ts [hunk 10]
```diff
@@ -7,54 +7,69 @@ export interface User {
...
+        let existingUser = this.users.find(existingUser => existingUser.id === id)
 
-    public updateUser = (id: number, user: User): User | null => {
-        console.log("Update user")
+        if (existingUser) {
+            if (existingUser.name === updatedUser.name) {
+                existingUser = { ...updatedUser }
+                return { id, ...existingUser }
+            } else 
+            throw new CustomError(500, "Different user with same id already exists.")
```
Comment (ianandrewson) [inline src/index.ts]: Clever.

**Review by ianandrewson — COMMENTED** [review]
Super solid. The only thing I would change are the status code if you want to, and not instantiating a new class in your create and delete methods.

File: src/index.ts [hunk 11]
```diff
@@ -7,54 +7,69 @@ export interface User {
...
-        ]
+    constructor(users: ReadonlyArray<User>) {
+        this.users = users
+    }
+
+    public getUserById = (id: number): User => {
+        const foundUser = this.users.find(user => user.id === id)
+        if (foundUser === undefined) {
+            throw new CustomError(500, "No user found.")
+        } else return foundUser
```
Comment (rachoac) [inline src/index.ts]: Braces

File: src/index.ts [hunk 12]
```diff
@@ -7,54 +7,69 @@ export interface User {
...
-                color: "red",
-            },
-        ]
+    constructor(users: ReadonlyArray<User>) {
+        this.users = users
+    }
+
+    public getUserById = (id: number): User => {
+        const foundUser = this.users.find(user => user.id === id)
+        if (foundUser === undefined) {
```
Comment (rachoac) [inline src/index.ts]: !foundUser

File: src/index.ts [hunk 13]
```diff
@@ -7,54 +7,69 @@ export interface User {
...
+    public createUser = (user: User): User => {
+        const userIdExists = this.users.find(existingUser => existingUser.id === user.id)
 
-    public createUser = (user: User): User | null => {
-        console.log("Create user", user)
-    }
+        if (userIdExists) {
+            throw new CustomError(500, "User with id already exists.")
+        } else { 
+            new UserAPI([user, ...this.users])
```
Comment (rachoac) [inline src/index.ts]: Why is this returning a new instance of the api

**Review by rachoac — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 14]
```diff
@@ -7,54 +7,69 @@ export interface User {
...
+    }  
 
-    public deleteUserById = (id: number): User | null => {
-        console.log("Create user by id")
-    }
+    public deleteUserById = (id: number): User => {
+        const user = this.users.find(existingUser => existingUser.id === id)
+
+        if (user) {
+            this.users = this.users.filter(user => user.id !== id)
```
Comment (ianandrewson) [inline src/index.ts:59]: Nice

File: src/index.ts [hunk 15]
```diff
@@ -7,54 +7,78 @@ export interface User {
...
-    }
+    public deleteUserById = (id: number): User => {
+        const user = this.users.find(existingUser => existingUser.id === id)
 
-    public updateUser = (id: number, user: User): User | null => {
-        console.log("Update user")
+        if (user) {
+            this.users = this.users.filter(user => user.id !== id)
+            return user
+        } else throw new CustomError(404, "No user with that id found.")
```
Comment (rachoac) [inline src/index.ts:61]: braces

File: src/index.ts [hunk 16]
```diff
@@ -7,54 +7,78 @@ export interface User {
...
+    } 
+
+    public updateUser = (id: number, updatedUser: User): User => {
+        let targetUser = this.users.find(existingUser => existingUser.id === id)
+
+        if (targetUser) {
+            this.users = this.users.map(user => {
+                    if (user.id === id) {
+                        if (user.name === updatedUser.name) {
```
Comment (rachoac) [inline src/index.ts:70]: cool
